expedition rewards

deltanedas opened the pull request:
## About the PR
requires changes to cargo system to be able to add and approve an item rather than a cargoProduct

Requires https://github.com/space-wizards/RobustToolbox/pull/4142

**Media**
rewards added upon ftling back, spawner needs to have sprite changed or something but yeah:
![16:14:58](https://github.com/space-wizards/space-station-14/assets/39013340/dec0b1e3-c7fc-4ffb-970a-290373eb3592)

spawner still spawns stuff:
![16:17:28](https://github.com/space-wizards/space-station-14/assets/39013340/e0450b8a-1e92-4d78-98bf-e7659206688d)

spawners use actual icons
![05:35:31](https://github.com/space-wizards/space-station-14/assets/39013340/f9d177f4-a6fc-4dba-8e49-a598292c1879)


- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**
:cl:
- add: NanoTrasen has finally started rewarding salvagers for actually completing their missions. Rewards can range from materials, to cloning parts, to advanced weaponry.
- remove: todo...

deltanedas commented:
@metalgearsloth besides already mentioned cargo concerns, how would i get the difficulty rating for the expedition component

metalgearsloth commented:
> @metalgearsloth besides already mentioned cargo concerns, how would i get the difficulty rating for the expedition component

You can probably put the difficulty on the comp itself if it's not there (rn you could just derive it from the seed but selectable difficulty means it'd need to be there).

metalgearsloth commented:
Add in cloner plz.

I think the amount of reward for 20 minutes work is probably too low, maybe make it spawn like 3 crates? Maybe like 3 common on normal, 1 common 2 rare on hard, 1 rare 2 epic on hardets or smth. I'd say 5 but if I'm leaning towards allowing multiple ops at once.

metalgearsloth commented on Content.Server/Salvage/SalvageSystem.Runner.cs:
```diff
@@ -272,8 +275,36 @@ private void UpdateRunner()
+            }
+        }
+
+        // give rewards if it was completed
+        var rewardQuery = EntityQueryEnumerator<SalvageExpeditionComponent>();
+        while (rewardQuery.MoveNext(out var uid, out var comp))
+        {
+            if (comp.Completed)
+            {
                 Announce(uid, Loc.GetString("salvage-expedition-completed"));
+                GiveReward(uid, comp);
             }
         }
     }
+
```
Isn't this going to spam the reward every tick?

deltanedas replied:
isnt it deleted after

deltanedas commented:
@metalgearsloth should the 3 rewards of different tiers be made apparent in mission ui or should that be left for when choosing mission difficulty is implemented

deltanedas commented:
@metalgearsloth ok rewards work (with inlined Frame0 to run duh) i got more questions:
- should i change the spawners sprite to be a t3/t4 part respectively so it looks correct in the cargo console or is there a smarter way that still looks like a spawner when mapping stuff
- should i remove random loot spawners from maps to not get 50 stacks of mats and not even complete the mission, or just make them rarer
- should i remove the laser cannon xray cannon etc from maps now that they are rewards

metalgearsloth commented:
> should i change the spawners sprite to be a t3/t4 part respectively so it looks correct in the cargo console or is there a smarter way that still looks like a spawner when mapping stuff
Uhhh yeah if it's not used anywhere.

> should i remove random loot spawners from maps to not get 50 stacks of mats and not even complete the mission, or just make them rarer

Steel sheets / canisters are fine to be mapped in I think, just anything rarer than that (e.g. xrays, lasers, ATV, APE) probably remove.

metalgearsloth commented:
So my general idea long-term:

- Explicit mission rewards are station shit like T3 parts or cloner or whatever.
- Missions themselves have either mining or common materials spawned, nothing rare, like basic circuitboards or whatever. At best it'd be something like a PKA upgrade or the likes from opposing salvagers.
- Expedition mining + shaft mining gives bulk minerals.
- Expeditions also have chemplants that can be ground up for when we remove the infinite chem dispenser but need sprites for this. Same with potential NPC critters.
- Salvage magnet pulls in wrecks you can deconstruct for money (using a similar UI to expeditions) and has no combat whatsoever and is procedurally generated (you'd pretty much re-use the dungeon code for this), although this stuff sells at a reduced rate. Think like hardspace shipbreakers.
- Shaft mining would probably just be nearby asteroids that are essentially what the cave mining missions for expeditions are like with very sporadic mobs.

deltanedas commented:
for last 2 basically afterlight :trollface:

metalgearsloth commented:
The last one would probably be using AL's worldgen though I have NFI how to use it.

deltanedas commented:
copy paste le asteroid generator but instead of procedurally for the entire map for a single position

metalgearsloth commented:
It's already in the repo but it has a billion options and its own queuedel implementation and scary.

deltanedas commented:
ok so:
- removed all mats and mat spawners
- removed akms gatfruit secfab lasers drozd revolvers and mags
- added some plushies and books in place of gold/silver/secfab

should i add gatfruit back as a high tier reward since funny

metalgearsloth commented:
> should i add gatfruit back as a high tier reward since funny

Sure

deltanedas commented:
,pro says after merging it :trollface:
